src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,143 @@
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Composition;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.MSBuild.BuildHostProcessManager;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+internal sealed class FileBasedProgramsProjectSystem : LanguageServerProjectLoader, ILspMiscellaneousFilesWorkspaceProvider
+{
+    private readonly ILspServices _lspServices;
+    private readonly ILogger<FileBasedProgramsProjectSystem> _logger;
+    private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+    public FileBasedProgramsProjectSystem(
+        ILspServices lspServices,
+        IMetadataAsSourceFileService metadataAsSourceFileService,
+        LanguageServerWorkspaceFactory workspaceFactory,
+        IFileChangeWatcher fileChangeWatcher,
+        IGlobalOptionService globalOptionService,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        ProjectLoadTelemetryReporter projectLoadTelemetry,
+        ServerConfigurationFactory serverConfigurationFactory,
+        BinlogNamer binlogNamer)
+            : base(
+                workspaceFactory.FileBasedProgramsProjectFactory,
+                workspaceFactory.TargetFrameworkManager,
+                workspaceFactory.ProjectSystemHostInfo,
+                fileChangeWatcher,
+                globalOptionService,
+                loggerFactory,
+                listenerProvider,
+                projectLoadTelemetry,
+                serverConfigurationFactory,
+                binlogNamer)
+    {
+        _lspServices = lspServices;
+        _logger = loggerFactory.CreateLogger<FileBasedProgramsProjectSystem>();
+        _metadataAsSourceFileService = metadataAsSourceFileService;
+    }
+
+    public Workspace Workspace => ProjectFactory.Workspace;
+
+    public async Task<TextDocument?> AddMiscellaneousDocumentAsync(DocumentUri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentFilePath = uri.ParsedUri is not null ? ProtocolConversions.GetDocumentFilePathFromUri(uri.ParsedUri) : uri.UriString;
+
+        // https://github.com/dotnet/roslyn/issues/78421: MAS should be its own workspace
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, documentText.Container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var document = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return document;
+        }
+
+        var primordialDoc = AddPrimordialDocument(uri, documentText, languageId);
+
+        if (uri.ParsedUri?.IsFile is true
+            && primordialDoc.Project.Language == LanguageNames.CSharp
```

Comment:
i would remove this check.  i don't know hwy we're special casing C# here.

Reply:
We aren't introducing the ability to have virtual project and do a design-time build on a loose file, except in the case of C#. So one thing we could do here is delegate this question to the 'VirtualProject' type. Simply ask, is this something you can make a virtual project for. Would that be preferable?

Reply:
If we removed this check, we would basically create a C# project for whatever non-C# file we are being asked about and run a design time build on it. And a bunch of unwanted diagnostics etc would result, I think.

Reply:
yes.  our design philosophy is:

1. if you have lang specific work, you either do it in a lang specific package (like MS.CA.Workspaces.CSharp), or
2. you defer to some service and allow it to exist or not (or have default behavior or not) for specific languages.  Like `ILanguageService`.

We do not have the core layers have special C#/VB/razor/etc. codepaths :)

Reply:
>  And a bunch of unwanted diagnostics etc would result, I think.

i would expect at some point something would go: this isn't something i can make a VP for and just bail out.  i'm not sure which component is best for that.  will defer to Jason here :)

Reply:
Frankly, I'm OK with a language name check here for now until we have a second language to figure out how to support. I imagine Razor will be that in the nearish future. Once we know what that'll look like more, it'll be clearer how to refactor.

Also practically we don't really have a good place to put it. We don't have language-specific language server projects at all. So we could do at this point is stick a random interface just to return "true" for C# at some lower layer, even though it only existed for the higher layer to consume. That feels dangerously close to https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition....

> We do not have the core layers have special C#/VB/razor/etc. codepaths :)

git grep disagrees here...I see quite a bunch. :smile:

Reply:
Also for:

> this isn't something i can make a VP for and just bail out

That might actually be the dotnet run-api side where it'd say it doesn't support the file.

Reply:
Note that right now the generation of the virtual project file also has a bunch of C# specific logic that would be going away with run-api too.